node/src/types/appendable_block.rs

```diff
@@ -0,0 +1,121 @@
+use datasize::DataSize;
+use num_traits::Zero;
+use thiserror::Error;
+
+use crate::{
+    components::block_proposer::DeployType,
+    types::{chainspec::DeployConfig, DeployHash, ProtoBlock, Timestamp},
+};
+
+#[derive(Debug, Error)]
+pub(crate) enum AddError {
+    #[error("would exceed maximum transfer count per block")]
+    TransferCount,
+    #[error("would exceed maximum deploy count per block")]
+    DeployCount,
+    #[error("would exceed maximum gas per block")]
+    GasLimit,
+    #[error("would exceed maximum block size")]
+    BlockSize,
+    #[error("duplicate deploy")]
+    Duplicate,
+    #[error("payment amount could not be converted to gas")]
+    InvalidGasAmount,
+    #[error("deploy is not valid in this context")]
+    InvalidDeploy,
+}
+
+/// A block that is still being added to. It keeps track of and enforces block limits.
+#[derive(Clone, DataSize, Debug)]
+pub struct AppendableBlock {
+    deploy_config: DeployConfig,
+    deploy_hashes: Vec<DeployHash>,
+    transfer_hashes: Vec<DeployHash>,
+    deploy_and_transfer_set: HashSet<DeployHash>,
+    timestamp: Timestamp,
+    #[data_size(skip)]
+    total_gas: Gas,
+    total_size: usize,
+}
+
+impl AppendableBlock {
+    /// Creates an empty `AppendableBlock`.
+    pub(crate) fn new(deploy_config: DeployConfig, timestamp: Timestamp) -> Self {
+        AppendableBlock {
+            deploy_config,
+            deploy_hashes: Vec::new(),
+            transfer_hashes: Vec::new(),
+            timestamp,
+            deploy_and_transfer_set: HashSet::new(),
+            total_gas: Gas::zero(),
+            total_size: 0,
+        }
+    }
+
+    /// Returns the total size of all deploys so far.
+    pub(crate) fn total_size(&self) -> usize {
+        self.total_size
+    }
+
+    /// Attempts to add a deploy to the block; returns an error if that would violate a validity
+    /// condition.
+    pub(crate) fn add(
+        &mut self,
+        hash: DeployHash,
+        deploy_type: &DeployType,
+    ) -> Result<(), AddError> {
+        if self.deploy_and_transfer_set.contains(&hash) {
+            return Err(AddError::Duplicate);
+        }
+        if !deploy_type
+            .header()
+            .is_valid(&self.deploy_config, self.timestamp)
+        {
+            return Err(AddError::InvalidDeploy);
+        }
+        if deploy_type.is_transfer() {
+            if self.transfer_hashes.len() == self.deploy_config.block_max_transfer_count as usize {
+                return Err(AddError::TransferCount);
+            }
+            self.transfer_hashes.push(hash);
+        } else {
+            if self.deploy_hashes.len() == self.deploy_config.block_max_deploy_count as usize {
+                return Err(AddError::DeployCount);
+            }
+            // Only deploys count towards the size and gas limits.
+            if self.total_size
+                > (self.deploy_config.max_block_size as usize).saturating_sub(deploy_type.size())
```

Comment:
I don't understand why you prefer it this way :sweat_smile: To me it's just much simpler to see the deploy's size added to the `self.total_size`.

Reply:
Right… trying to cover all the overflow/underflow and `usize::MAX` cases. But maybe I can use a `checked_add`. :+1:

Reply:
Doen in https://github.com/CasperLabs/casper-node/pull/1276/commits/0994e702c9eba2795f6cd3d669c9398b925b213c.